### DISCUSSION TITLE
How to write the edited .json back to the .json file

### DISCUSSION THREAD

**Discussion by Basilisvirus** [nlohmann/json#3737] [Q&A] [answered]
I open the .json file, I edit it. Then, I want to write it back to the .json file. How do I do that?

Here is what I tried:

```
            std::cout<<"run once\n\r";
            std::fstream myfile;
            myfile.open("members.json");

            json membersJson = json::parse(myfile);
            std::cout << "Before: " << membersJson << '\n';
            membersJson["pi"] = 3.141;
            std::cout << "After: " << membersJson << '\n';
            std::string s = membersJson.dump(4);

            myfile << s;

            myfile.close();
```

The output is:

![image](https://user-images.githubusercontent.com/18421898/189499797-0ae55f54-e436-4652-bd16-07475ac90cfb.png)

But the .json file does not change. It remains with its old content: {} .

I have tried without the .dump() function, so the code was:

```
            std::cout<<"run once\n\r";
            std::fstream myfile;
            myfile.open("members.json");

            json membersJson = json::parse(myfile);
            std::cout << "Before: " << membersJson << '\n';
            membersJson["pi"] = 3.141;
            std::cout << "After: " << membersJson << '\n';

            myfile << membersJson ;

            myfile.close();
```

But still the file remains the same, unchanged.

I also tried using `std::stream` instead of std::ifstream , since I think `std::ifstream` is only for reading files and not writing. But no luck.

**Comment by falbrechtskirchinger** [comment]
Your problem has nothing to do with this library.

Your usage of `std::fstream` is incorrect. People have been struggling with reading and writing files for decades. Google it and you'll drown in the number of answers. I won't add to that list.

**Reply by Basilisvirus** [reply]
Hm thanks, @ichenglin suggested I close the file, and then re-open it, to write the contents back in, which works.
I didn't thought I had to close and re-open the file, since I do not have to do that in regular read-write procedures with `std::fstream`.

When I open a file with `fstream `and I then use` json membersJson = json::parse(input_file);` , the file cannot be written to anymore, I think because the json library is taking control of it. Isn't it something that should be mentioned in the documents? without the `json::parse` , the file can then be written as normal. 

But I didnt find anyone having the same problem with me, so I guess I didnt gave it enough time to look it up.

**Reply by gregmarr** [reply]
> I didn't thought I had to close and re-open the file, since I do not have to do that in regular read-write procedures with std::fstream.

If you read the file, then before overwriting it, you need to reset to the beginning or close and re-open.

>file cannot be written to anymore, I think because the json library is taking control of it. 

It doesn't "take control" of the file, it reads it.  It's no different than opening the file and reading it yourself.

**Comment by ichenglin** [comment, marked as answer]
There are several ways of **replace the content** of the **same file**,
such as **clearing the end-of-file** error,
then **relocating the output pointer** to the beginning of the file to **replace the original content**.

However apart from all the solutions,
the **simplest way** is just to **separate the read/write file process**,
such as the example I provided below.

```cpp
// read file content
std::ifstream input_file("members.json");
nlohmann::json file_json = nlohmann::json::parse(input_file);
input_file.close();
// adjust json
file_json["pi"] = 3.141;
// write file content
std::ofstream output_file("members.json");
std::string file_json_string = file_json.dump();
output_file.write(file_json_string.c_str(), file_json_string.size());
output_file.close();
```

**Reply by Basilisvirus** [reply]
Thanks! I didnt think I had to close and re-open the file, as this is not needed when I use `fstream `without this json library. I think an example like this should exist in the readme file. It doesn't take much space.

**Reply by ichenglin** [reply]
You actually **don't have to close the file** to write to it in **`fstream`**,
but I find the **solution easier** than any other else so I posted it.
